File: Java-Core/src/main/java/prompto/literal/RangeLiteral.java
package prompto.literal;

import prompto.error.PromptoError;
import prompto.error.SyntaxError;
import prompto.expression.IExpression;
import prompto.runtime.Context;
import prompto.type.IType;
import prompto.type.IntegerType;
import prompto.utils.CodeWriter;
import prompto.value.IValue;

public class RangeLiteral implements IExpression {

	IExpression first;
	IExpression last;
	
	public RangeLiteral(IExpression first, IExpression last) {
		this.first = first;
		this.last = last;
	}

	@Override
	public String toString() {
		return "[" + first.toString() + ".." + last.toString() + "]";
	}
	
	@Override
	public void toDialect(CodeWriter writer) {
		writer.append("[");
		first.toDialect(writer);
		writer.append("..");
		last.toDialect(writer);
		writer.append("]");
	}

	public IExpression getFirst() {
		return first;
	}
	
	public IExpression getLast() {
		return last;
	}
	
	@Override
	public IType check(Context context) throws SyntaxError {
		IType firstType = first.check(context);
		IType lastType = last.check(context);
		return firstType.checkRange(context,lastType);
	}
	
	@Override
	public IValue interpret(Context context) throws PromptoError {
		IType type = first.check(context);
		if("IntegerLimits".equals(type.getName()))
			type = IntegerType.instance();
		Object of = first.interpret(context);
		Object ol = last.interpret(context);
		return type.newRange(of,ol);
	}

}

File: Java-CodeServer/src/main/resources/js/prompto/literal/RangeLiteral.js
function RangeLiteral(first, last) {
	this.first = first;
	this.last = last;
	return this;
}


RangeLiteral.prototype.toString = function() {
	return "[" + this.first.toString() + ".." + this.last.toString() + "]";
};

RangeLiteral.prototype.toDialect = function(writer) {
    writer.append("[");
    this.first.toDialect(writer);
    writer.append("..");
    this.last.toDialect(writer);
    writer.append("]");
};


RangeLiteral.prototype.check = function(context) {
	var firstType = this.first.check(context);
	var lastType = this.last.check(context);
	return firstType.checkRange(context,lastType);
};

RangeLiteral.prototype.interpret = function(context) {
	var type = this.first.check(context);
	if("IntegerLimits"==type.name) {
		type = IntegerType.instance;
	}
	var of = this.first.interpret(context);
	var ol = this.last.interpret(context);
	return type.newRange(of,ol);
};


exports.RangeLiteral = RangeLiteral;

